### PR TITLE
[SYCLomatic] Refactor the math API rewriter: using tag to hint the dispatch

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/CallExprRewriterMath.cpp [hunk 1]
```diff
@@ -931,6 +931,190 @@ auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+enum class Tag {
+  _0 = 0,
+  _1,
+  _2,
+  _3,
+  _4,
+  _5,
+  _6,
+  _7,
+  _8,
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriterMath.cpp]: remove

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 2]
```diff
@@ -238,6 +238,183 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+                        std::pair<std::string,
+                                  std::shared_ptr<CallExprRewriterFactoryBase>>,
+                        int>> &MathAPIRewriters) {
+  return std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>(
+      Name, std::make_shared<MathRewriterFactory>(Name, MathAPIRewriters));
+}
+
+template <class T>
+inline std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>
+createMathRewriterFactory(
```
Comment (zhiweij1) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]: use args...

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 3]
```diff
@@ -238,6 +238,183 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+  unsupported_warning,
+  no_rewrite = 999
+};
+
+inline std::function<bool(const CallExpr *)> TrueFunctor =
+    [](const CallExpr *) { return true; };
+
+class MathRewriterFactory final : public CallExprRewriterFactoryBase {
+  std::string Name;
+  std::unordered_map<
```
Comment (zhiweij1) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]: use array

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 4]
```diff
@@ -238,6 +238,183 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+      Tag, std::tuple<std::function<bool(const CallExpr *)>,
```
Comment (zhiweij1) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]: use std::pair

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 5]
```diff
@@ -238,6 +239,167 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
 };
 } // namespace math
 
+enum class Tag : size_t {
+  device_perf = 0,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]: Please add comment for the meaning of each ENUM value.

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 6]
```diff
@@ -238,6 +239,167 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+  element_t &HostPerfRewriter =
+      MathAPIRewriters[static_cast<size_t>(Tag::host_perf)];
+  element_t &HostNormalRewriter =
+      MathAPIRewriters[static_cast<size_t>(Tag::host_normal)];
+  element_t &UnsupportedWarningRewriter =
+      MathAPIRewriters[static_cast<size_t>(Tag::unsupported_warning)];
+  element_t &NoRewriteRewriter =
+      MathAPIRewriters[static_cast<size_t>(Tag::no_rewrite)];
+
+  std::shared_ptr<CallExprRewriter> getDeviceRewriter(const CallExpr *C) const {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]: Please add a comment for the priority.

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 7]
```diff
@@ -238,6 +239,167 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+                      const array_t &MathAPIRewritersInput)
+      : Name(Name), MathAPIRewriters(MathAPIRewritersInput) {
+    NoRewriteRewriter = std::make_pair(
+        TrueFunctor,
+        std::make_pair(Name,
+                       std::dynamic_pointer_cast<CallExprRewriterFactoryBase>(
+                           std::make_shared<NoRewriteFuncNameRewriterFactory>(
+                               Name, Name))));
+  }
+  std::shared_ptr<CallExprRewriter> create(const CallExpr *C) const override {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h:342]: Please add comment for the priority decision.

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 8]
```diff
@@ -238,6 +239,167 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
 };
 } // namespace math
 
+enum class Tag : size_t {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h:241]: if only used by math, may move to math:: namespace or rename to math..tag.

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 9]
```diff
@@ -236,8 +237,181 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
+
+template <typename... Ts>
+inline void createMathRewriterFactoryImpl(
+    MathRewriterFactory::array_t &Rewriters, math::Tag Tag,
+    std::function<bool(const CallExpr *)> Cond,
+    std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>
+        Rewriter,
+    int, Ts... Args) {
+  createMathRewriterFactoryImpl(Rewriters, Args...);
+  Rewriters[static_cast<size_t>(Tag)] = std::make_pair(Cond, Rewriter);
```
Comment (tangjj11) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h:400]: Please add a duplicated check and report error.

**Review by tangjj11 — CHANGES_REQUESTED** [review]

**Review by tangjj11 — APPROVED** [review]

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 10]
```diff
@@ -236,8 +237,200 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
...
 };
+
+enum class Tag : size_t {
+  device_perf = 0,  // device API for performace
+  device_normal,    // device API
+  math_libdevice,   // device API using libdevice
+  device_std,       // device API using std namespace
+  emulation,        // emulation
+  ext_experimental, // device API using experimental feature
+  host_perf,        // host API for performace
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]:
```suggestion
  host_perf,        // host API for performance
```

File: clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h [hunk 11]
```diff
@@ -236,8 +237,200 @@ inline auto UsingDpctMinMax = [](const CallExpr *C) -> bool {
   Arg1T.removeLocalFastQualifiers(Qualifiers::CVRMask);
   return Arg0T != Arg1T;
 };
+
+enum class Tag : size_t {
+  device_perf = 0,  // device API for performace
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rewriters/Math/CallExprRewriterMath.h]:
```suggestion
  device_perf = 0,  // device API for performance
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**